domain/blockdevice/state/state.go

```diff
@@ -0,0 +1,401 @@
+			return errors.Trace(err)
+		}
+		blockDevicesByUUID[id] = bd
+		dbBlockDevice := BlockDevice{
+			ID:             id.String(),
+			DeviceName:     bd.DeviceName,
+			Label:          bd.Label,
+			DeviceUUID:     bd.UUID,
+			HardwareId:     bd.HardwareId,
+			WWN:            bd.WWN,
+			BusAddress:     bd.BusAddress,
+			SerialId:       bd.SerialId,
+			MountPoint:     bd.MountPoint,
+			Size:           bd.Size,
+			FilesystemType: fsTypeID,
+			InUse:          bd.InUse,
+		}
+		if err := tx.Query(ctx, insertStmt, dbBlockDevice).Run(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	insertLinkQuery := `
+INSERT INTO block_device_link_device (block_device_uuid, name)
+VALUES (
+    $DeviceLink.block_device_uuid,
+    $DeviceLink.name
+)
+`
+	insertStmt, err = sqlair.Prepare(insertLinkQuery, DeviceLink{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for uuid, bd := range blockDevicesByUUID {
+		for _, link := range bd.DeviceLinks {
+			dbDeviceLink := DeviceLink{
+				ParentUUID: uuid.String(),
+				Name:       link,
+			}
+			if err := tx.Query(ctx, insertStmt, dbDeviceLink).Run(); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+
+	insertJoinQuery := `
+INSERT INTO block_device_machine(block_device_uuid, machine_uuid)
+VALUES ($M.block_device_uuid, $M.machine_uuid)
+`
+	insertJoinStmt, err := sqlair.Prepare(insertJoinQuery, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for uuid := range blockDevicesByUUID {
+		if err := tx.Query(ctx, insertJoinStmt, sqlair.M{
+			"machine_uuid":      machineUUID,
+			"block_device_uuid": uuid.String(),
+		}).Run(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
+func blockDevicesChanged(oldDevices, newDevices []blockdevice.BlockDevice) bool {
+	if len(oldDevices) != len(newDevices) {
+		return true
+	}
+	for _, o := range oldDevices {
+		var found bool
+		for _, n := range newDevices {
+			if reflect.DeepEqual(o, n) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return true
+		}
+	}
+	return false
+}
+
+// RemoveMachineBlockDevices removes all the block devices for the specified machine.
+// It is the same as calling SetMachineBlockDevices with an empty list, but does not
+// error if the machine life is Dead.
+func (st *State) RemoveMachineBlockDevices(ctx context.Context, machineId string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		q := `
+SELECT machine.uuid AS &M.machine_uuid
+FROM   machine
+WHERE  machine.machine_id = $M.machine_id
+`
+		stmt, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		result := sqlair.M{}
+		err = tx.Query(ctx, stmt, sqlair.M{"machine_id": machineId}).Get(result)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Trace(err)
+		}
+		if len(result) == 0 {
+			return fmt.Errorf("machine %q %w", machineId, errors.NotFound)
```

Comment:
Because this is being used by other operations in state. Specifically big ones like block devices and is going to get surfaced via those functions where lots of operations are happening we have no way to know what wasn't found. Especially when we are dealing with fs type. We need to make this a typed error for Machine Not Found

Reply:
Yes, yes we do. It is coming....

Reply:
I've relented and bloated this PR a little more by doing it now instead of in the next PR.